pythonforandroid/build.py

```diff
@@ -709,25 +703,157 @@ def run_pymodules_install(ctx, modules):
             "site-packages")) + ":" + env["PYTHONPATH"]
-        shprint(sh.bash, '-c', (
-            "source venv/bin/activate && " +
-            "pip install -v --target '{0}' --no-deps -r requirements.txt"
-        ).format(ctx.get_site_packages_dir().replace("'", "'\"'\"'")),
-                _env=copy.copy(env))
+
+        # Install the manually specified requirements first:
+        if not modules:
+            info('There are no Python modules to install, skipping')
+        else:
+            info('Creating a requirements.txt file for the Python modules')
+            with open('requirements.txt', 'w') as fileh:
+                for module in modules:
+                    key = 'VERSION_' + module
+                    if key in environ:
+                        line = '{}=={}\n'.format(module, environ[key])
+                    else:
+                        line = '{}\n'.format(module)
+                    fileh.write(line)
+
+            info('Installing Python modules with pip')
+            info('IF THIS FAILS, THE MODULES MAY NEED A RECIPE. '
+                 'A reason for this is often modules compiling '
+                 'native code that is unaware of Android cross-compilation '
+                 'and does not work without additional '
+                 'changes / workarounds.')
+
+            shprint(sh.bash, '-c', (
+                "source venv/bin/activate && " +
+                "pip install -v --target '{0}' --no-deps -r requirements.txt"
+            ).format(ctx.get_site_packages_dir().replace("'", "'\"'\"'")),
+                    _env=copy.copy(env))
+
+        # Afterwards, run setup.py if present:
+        if (project_dir is not None and (
+                    os.path.exists(os.path.join(project_dir, "setup.py")) and
+                    not ignore_setup_py
+                    )):
+            with current_directory(project_dir):
+                info('setup.py is present, running project install. ' +
+                     '(disable this behavior with --ignore-setup-py)')
+
+                # Output some debug info:
+                info('Contents that will be used for constraints.txt:')
+                shprint(sh.bash, '-c', (
+                    "source '" + join(
+                        ctx.build_dir, "venv", "bin", "activate"
+                    ).replace("'", "'\"'\"'") + "' && " +
+                    "pip freeze"
+                ), _env=copy.copy(env))
+
+                # Make sure all packages found are fixed in version
+                # by writing a constraint file, to avoid recipes being
+                # upgraded & reinstalled:
+                with open('constraints.txt', 'wb') as fileh:
+                    output = subprocess.check_output([
+                        "bash", "-c",
+                        "source '" + join(
+                            ctx.build_dir, "venv", "bin", "activate"
+                        ).replace("'", "'\"'\"'") + "' && " +
+                        "pip freeze"
```

Comment:
Minor DRY: can't we run that command once and save the output, then once we print it, and the other time we write it to file.
Or simply you write it to that `constraints.txt` file as you did and then you just print the file

Reply:
hopefully I remember this correctly, but I do have a `constraints` variable now so I think this should be covered! if not poke me, but for now it seems like this is resolved :+1: